rustdoc: Merge source code pages HTML elements together

GuillaumeGomez opened the pull request:
We realized that the HTML generated for the source code pages could be improved quite a lot. This PR is a first pass toward this goal. Some explanations: it merges similar classes elements (even when there are white characters in between).

There is an exception to this: if this is an ident, I also merged it with "unclassified" elements. This part is up to debate and can be very easily removed as the check is performed in one place (in the `can_merge` function).

EDIT: The `ident` is now only kept in the code for the `span` it contains but it is not rendered into the HTML.

So now some numbers:

For these ones, on each page, I run this JS: `document.getElementsByTagName('*').length`. The goal is to count the number of DOM elements. I took some pages that seemed big, but don't hesitate to check some others. 

| file name | before this PR | with this PR | diff | without ident | diff |
|-|-|-|-|-|-|
| std/lib.rs.html (source link on std crate page) | 3455 | 2776 | 20.7% | 2387 | 31% |
| alloc/vec/mod.rs.html (source on Vec type page) | 11012 | 8084 | 26.6% | 6682 | 39.4% |
| alloc/string.rs.html (source on String type page) | 10800 | 8214 | 24% | 6712 | 37.9% |
| std/sync/mutex.rs.html (source on Mutex type page) | 2953 | 2403 | 18.7% | 2139 | 27.6% |

You can test it [here](https://rustdoc.crud.net/imperio/merge-html-elements-together/src/std/lib.rs.html).

cc @jsha 
r? @notriddle

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7796330045?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/1271f1db650c3f5c57c46678cbf29941c21af8a2/checks/7796330045/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustdoc/html/highlight.rs at line 220:
         };
     });
-    write_pending_elems(
-        out,
-        out,
-        &href_context,
-        &mut pending_elems,
-        &mut current_class,
-        &closing_tags,
-    );
+    write_pending_elems(out, &href_context, &mut pending_elems, &mut current_class, &closing_tags);
 
 
 fn write_footer(out: &mut Buffer, playground_button: Option<&str>) {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/html/tests.rs" "/checkout/src/librustdoc/html/render/search_index.rs" "/checkout/src/librustdoc/html/markdown/tests.rs" "/checkout/src/librustdoc/html/static_files.rs" "/checkout/src/librustdoc/html/format.rs" "/checkout/src/librustdoc/html/highlight.rs" "/checkout/src/librustdoc/html/highlight/tests.rs" "/checkout/src/librustdoc/html/render/context.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

GuillaumeGomez commented:
> Is this blocked on #100409, or does it replace that PR?

It's a separate one, but I'll need to fix conflicts on the tests if it is merged first.

GuillaumeGomez commented:
Fixed the format issue.

GuillaumeGomez commented on src/librustdoc/html/highlight/fixtures/union.html:
```diff
@@ -1,8 +1,8 @@
+<span class="kw">union </span><span class="ident">Foo {
+    i: i8,
+    u: i8,
 }
```
It's the part up to debate since most of the gain comes from being very lax on `ident`. I kinda like having the "ident" class since it still provides an information, but maybe we could completely get rid of it. Not sure.

An opinion maybe @jsha?

jsha replied:
Technically we have a style that matches .ident but it's pretty narrowly scoped:

```
pre.rust .attribute, pre.rust .attribute .ident { color: #C82829; }
```

I agree with @notriddle: we're not using the `ident` class for anything right now, so let's get rid of it.

jsha commented:
Are there some unrelated changes in here related to type aliases?

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -102,6 +101,7 @@ pub(crate) fn eval_nullary_intrinsic<'tcx>(
             | ty::Never
             | ty::Tuple(_)
             | ty::Error(_) => ConstValue::from_machine_usize(0u64, &tcx),
+            ty::TyAlias(..) => bug!("unexpected TyAlias in eval_nullary_intrinsic"),
```
Why are there interpreter changes in a PR that supposedly only does something to the rustdoc HTML  output...?

GuillaumeGomez replied:
O.o

I pushed on the wrong branch...

GuillaumeGomez commented:
Yes completely, I was working on something else and pushed to the wrong branch. It is now fixed.

GuillaumeGomez commented:
I updated the online demo, added the new numbers. So in short: we need to keep `Class::Ident` around because we need the `Span` but I don't render it into the HTML anymore. The difference in the DOM size is quite big.

bors commented:
:umbrella: The latest upstream changes (presumably #100456) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Fixed the merge conflict with the `op` class removal.

bors commented:
:pushpin: Commit 8ed674c0f5161ff189d8a073a405ba8610056545 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 8ed674c0f5161ff189d8a073a405ba8610056545 -->
<!-- homu: {"type":"Approved","sha":"8ed674c0f5161ff189d8a073a405ba8610056545","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #100511) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7825489550?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6caff4ba3bb91053443320c8e8b86e338caeb58f/checks/7825489550/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [rustdoc] src/test/rustdoc/macro_rules-matchers.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/macro_rules-matchers" "/checkout/src/test/rustdoc/macro_rules-matchers.rs"
stdout: none
--- stderr -------------------------------
8: Invalid number of @has arguments
 // @has - ' todo {'
Encountered 1 errors
------------------------------------------


```

</details>

GuillaumeGomez commented:
@bors r=notriddle

bors commented:
:pushpin: Commit 6e574e100c47d1800e05d14cbaae959c11f320ed has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 6e574e100c47d1800e05d14cbaae959c11f320ed -->
<!-- homu: {"type":"Approved","sha":"6e574e100c47d1800e05d14cbaae959c11f320ed","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6e574e100c47d1800e05d14cbaae959c11f320ed with merge 801821d1560f84e4716fcbd9244ec959320a13d5...
<!-- homu: {"type":"BuildStarted","head_sha":"6e574e100c47d1800e05d14cbaae959c11f320ed","merge_sha":"801821d1560f84e4716fcbd9244ec959320a13d5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7827894176?check_suite_focus=true)
Approved by: notriddle
Pushing 801821d1560f84e4716fcbd9244ec959320a13d5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"notriddle","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7827894176?check_suite_focus=true"},"merge_sha":"801821d1560f84e4716fcbd9244ec959320a13d5"} -->

rust-timer commented:
Finished benchmarking commit (801821d1560f84e4716fcbd9244ec959320a13d5): [comparison url](https://perf.rust-lang.org/compare.html?start=4c5665583815a0f0f3e22516441efb43ea6dede2&end=801821d1560f84e4716fcbd9244ec959320a13d5&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4c5665583815a0f0f3e22516441efb43ea6dede2&end=801821d1560f84e4716fcbd9244ec959320a13d5&stat=instructions:u)

- Primary benchmarks: ❌ relevant regressions found
- Secondary benchmarks: ❌ relevant regressions found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 0.6%     | 1.4% | 12        |
| Regressions ❌ <br /> (secondary) | 1.8%     | 3.1% | 8         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -        | -   | 0         |
| All ❌✅ (primary) | 0.6%     | 1.4% | 12        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4c5665583815a0f0f3e22516441efb43ea6dede2&end=801821d1560f84e4716fcbd9244ec959320a13d5&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -   | 0         |
| Regressions ❌ <br /> (secondary) | 1.5%     | 1.5% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -2.3%    | -2.3% | 1         |
| All ❌✅ (primary) | -        | -   | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4c5665583815a0f0f3e22516441efb43ea6dede2&end=801821d1560f84e4716fcbd9244ec959320a13d5&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: ❌ relevant regression found
- Secondary benchmarks: ❌ relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions ❌ <br /> (primary) | 3.7%     | 3.7% | 1         |
| Regressions ❌ <br /> (secondary) | 3.1%     | 3.1% | 1         |
| Improvements ✅ <br /> (primary) | -        | -   | 0         |
| Improvements ✅ <br /> (secondary) | -        | -   | 0         |
| All ❌✅ (primary) | 3.7%     | 3.7% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance
